File: app/src/main/java/com/sdk/billinglibrary/BillingManager.java
package com.sdk.billinglibrary;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.android.billingclient.api.AcknowledgePurchaseParams;
import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.ProductDetails;
import com.android.billingclient.api.ProductDetailsResponseListener;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchasesResponseListener;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.android.billingclient.api.QueryProductDetailsParams;
import com.android.billingclient.api.QueryPurchasesParams;
import com.android.billingclient.api.SkuDetails;
import com.android.billingclient.api.SkuDetailsParams;
import com.sdk.billinglibrary.interfaces.IOnInitializationComplete;
import com.sdk.billinglibrary.interfaces.IOnPurchaseListener;
import com.sdk.billinglibrary.interfaces.ISkuListener;

import java.util.ArrayList;
import java.util.List;

class BillingManager implements BillingClientStateListener,
        PurchasesUpdatedListener, PurchasesResponseListener {

    static final String LOG_TAG = "MYTAG (Billing)";

    private static Context mContext;
    private static BillingManager mManager;

    static void initialize(@NonNull Context context, IOnInitializationComplete listener) {
        LocalConfig.init(context);
        mContext = context.getApplicationContext();
        if (mManager == null)
            mManager = new BillingManager(listener);
    }

    static BillingManager getInstance() {
        return mManager;
    }

    private BillingClient mBillingClient;
    private IOnPurchaseListener mOnPurchaseListener;
    private final IOnInitializationComplete mOnInitializationListener;

    private BillingManager(IOnInitializationComplete listener) {
        mOnInitializationListener = listener;
        restart();
    }

    private void restart() {
        if (mBillingClient != null) {
            mBillingClient.endConnection();
            mBillingClient = null;
        }
        mBillingClient = BillingClient
                .newBuilder(mContext)
                .setListener(this)
                .enablePendingPurchases()
                .build();
        mBillingClient.startConnection(this);
    }

    @Override
    public void onBillingSetupFinished(@NonNull BillingResult billingResult) {
        if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
            Log.d(LOG_TAG, "Billing setup finished!");

            boolean isSubSupported = isSubscriptionSupported();

            if (isSubSupported) {
                Log.d(LOG_TAG,"Subscriptions are supported!");
                queryPurchases();
            } else {
                Log.d(LOG_TAG, "Subscription are not supported!");
                LocalConfig.subscribeLocally(true);
                mOnInitializationListener.onComplete();
            }

        } else {
            Log.d(LOG_TAG, "Billing setup failed: "
                    + billingResult.getResponseCode() + " | "
                    + billingResult.getDebugMessage());
            mOnInitializationListener.onComplete();
        }
    }

    @Override
    public void onBillingServiceDisconnected() {
        restart();
    }

    @Override
    public void onQueryPurchasesResponse(@NonNull BillingResult billingResult, @NonNull List<Purchase> list) {
        if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
            acknowledgePurchases(list);
            Log.d(LOG_TAG, "Purchases are acknowledged!");
        } else {
            Log.d(LOG_TAG, "Failed to query purchases: "
                    + billingResult.getResponseCode() + " | "
                    + billingResult.getDebugMessage());
        }
        mOnInitializationListener.onComplete();
    }

    @Override
    public void onPurchasesUpdated(@NonNull BillingResult billingResult, @Nullable List<Purchase> list) {

        if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK
                && list != null) {

            boolean purchased = acknowledgePurchases(list);

            if (mOnPurchaseListener == null)
                return;

            if (purchased)
                mOnPurchaseListener.onPurchaseDone();
            else
                mOnPurchaseListener.onPurchaseCancelled();

        } else if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.USER_CANCELED) {
            if (mOnPurchaseListener != null)
                mOnPurchaseListener.onPurchaseCancelled();
        } else {
            if (mOnPurchaseListener != null)
                mOnPurchaseListener.onPurchaseFail();
        }
    }

    private boolean acknowledgePurchases(List<Purchase> list) {

        boolean purchased = false;

        for (Purchase purchase : list) {
            if (purchase.getPurchaseState() == Purchase.PurchaseState.PURCHASED) {
                if (!purchase.isAcknowledged()) {
                    AcknowledgePurchaseParams params = AcknowledgePurchaseParams.newBuilder()
                            .setPurchaseToken(purchase.getPurchaseToken())
                            .build();
                    mBillingClient.acknowledgePurchase(params, result -> {});
                }
                purchased = true;
            }
        }
        LocalConfig.subscribeLocally(purchased);
        return purchased;
    }

    private boolean isSubscriptionSupported() {
        return mBillingClient.isFeatureSupported(BillingClient.FeatureType.SUBSCRIPTIONS)
                .getResponseCode() == BillingClient.BillingResponseCode.OK;
    }

    void queryPurchases() {
        mBillingClient.queryPurchasesAsync(
                QueryPurchasesParams.newBuilder()
                        .setProductType(BillingClient.ProductType.SUBS)
                        .build(),
                this);
    }

    void retrieveSubs(String trialSubId, String premiumSubId, ISkuListener listener) {

        if (trialSubId == null
                || trialSubId.isEmpty()
                || premiumSubId == null
                || premiumSubId.isEmpty()
                || !mBillingClient.isReady()) {
            listener.onFailed();
            return;
        }

        SkuDetailsParams.Builder params = SkuDetailsParams.newBuilder();
        List<String> subs = new ArrayList<>();
        subs.add(trialSubId);
        subs.add(premiumSubId);
        params.setSkusList(subs).setType(BillingClient.SkuType.SUBS);

        List<QueryProductDetailsParams.Product> products = new ArrayList<>();
        QueryProductDetailsParams.Product.Builder builder = QueryProductDetailsParams.Product.newBuilder();
        products.add(builder.setProductId(trialSubId).setProductType(BillingClient.ProductType.SUBS).build());
        products.add(builder.setProductId(premiumSubId).setProductType(BillingClient.ProductType.SUBS).build());

        QueryProductDetailsParams queryProductDetailsParams =
                QueryProductDetailsParams.newBuilder()
                        .setProductList(products)
                        .build();

        mBillingClient.queryProductDetailsAsync(queryProductDetailsParams,
                (billingResult, productDetailsList) -> {

                    if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {

                        if (productDetailsList.size() != 2) {
                            Log.d(LOG_TAG, "Fetched too few SKUs, should be 2!");
                            listener.onFailed();
                        } else {

                            ProductDetails trial = null;
                            ProductDetails full = null;

                            for (ProductDetails product : productDetailsList) {
                                if (product.getProductId().equals(trialSubId))
                                    trial = product;
                                else if (product.getProductId().equals(premiumSubId))
                                    full = product;
                            }

                            if (trial != null && full != null)
                                listener.onResult(trial, full);
                            else
                                listener.onFailed();
                        }

                    } else {
                        listener.onFailed();
                    }
        });
    }

    void launchPurchaseFlow(Activity activity, ProductDetails product, String token, IOnPurchaseListener listener) {
        mOnPurchaseListener = listener;
        if (!mBillingClient.isReady() || product == null) {
            if (mOnPurchaseListener != null) {
                mOnPurchaseListener.onError();
            }
            return;
        }

        try {
            List<BillingFlowParams.ProductDetailsParams> params = new ArrayList<>();
            params.add(BillingFlowParams.ProductDetailsParams.newBuilder()
                            .setProductDetails(product)
                            .setOfferToken(token)
                    .build());
            BillingFlowParams flowParams = BillingFlowParams.newBuilder()
                            .setProductDetailsParamsList(params).build();
            mBillingClient.launchBillingFlow(activity, flowParams);
        } catch (IllegalArgumentException e) {
            mOnPurchaseListener.onError();
        }

    }

}

File: app/src/main/java/com/sdk/billinglibrary/BillingManager.kt
package com.sdk.billinglibrary

import android.app.Activity
import android.content.Context
import android.util.Log
import android.widget.Toast
import com.android.billingclient.api.AcknowledgePurchaseParams
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.BillingFlowParams.ProductDetailsParams
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.PendingPurchasesParams
import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import com.sdk.billinglibrary.LocalConfig.subscribeLocally

private val BANNED_CURRENCIES = listOf("INR", "MYR")

class BillingManager(
    val context: Context) {

    private val onPurchase: BillingPurchaseListener = object : BillingPurchaseListener {
        override fun onPurchaseDone() {
            Toast.makeText(context, R.string.purchase_done, Toast.LENGTH_LONG).show()
            Billing.listeners.forEach { it.onPurchaseDone() }
        }

        override fun onPurchaseFail() {
            Toast.makeText(context, R.string.purchase_fail, Toast.LENGTH_LONG).show()
            Billing.listeners.forEach { it.onPurchaseFail() }
        }

        override fun onPurchaseCancelled() {
            Billing.listeners.forEach { it.onPurchaseCancelled() }
        }

        override fun onError() {
            Toast.makeText(context, R.string.purchase_fail, Toast.LENGTH_LONG).show()
            Billing.listeners.forEach { it.onError() }
        }
    }

    private val client: BillingClient = BillingClient
        .newBuilder(context)
        .setListener { result, list ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK && list != null) {
                val purchased = acknowledgePurchases(list)
                if (purchased)
                    onPurchase.onPurchaseDone()
                else
                    onPurchase.onPurchaseCancelled()
            } else if (result.responseCode == BillingClient.BillingResponseCode.USER_CANCELED) {
                onPurchase.onPurchaseCancelled()
            } else {
                onPurchase.onPurchaseFail()
            }
        }
        .enablePendingPurchases(PendingPurchasesParams.newBuilder()
            .enableOneTimeProducts().build())
        .build()

    fun initialize(vararg productIds: String) {
        Log.d(Billing.LOG, "Initialization of billing manager...")
        client.startConnection(object : BillingClientStateListener {

            override fun onBillingServiceDisconnected() {
                // client.startConnection(this)
            }

            override fun onBillingSetupFinished(result: BillingResult) {

                if (Billing.isInitialized.value == true)
                    return

                if (result.responseCode == BillingClient.BillingResponseCode.OK)
                {
                    Log.d(Billing.LOG, "Billing setup finished!")
                    if (isSubscriptionSupported())
                    {
                        Log.d(Billing.LOG, "Subscriptions are supported!")
                        querySubsPurchases()
                        querySubsProducts(*productIds)
                    }
                    else
                    {
                        Log.d(Billing.LOG, "Subscription are not supported!")
                        subscribeLocally(Billing.UNSUPPORTED)
                        Billing.isInitialized.postValue(true)
                    }
                }
                else
                {
                    Log.d(Billing.LOG, "Billing setup failed: ${result.responseCode} | ${result.debugMessage}")
                    Billing.isInitialized.postValue(true)
                }
            }
        })
    }

    private fun isSubscriptionSupported(): Boolean {
        val code = client.isFeatureSupported(BillingClient.FeatureType.SUBSCRIPTIONS).responseCode
        return code == BillingClient.BillingResponseCode.OK
    }

    private fun querySubsProducts(vararg ids: String) {

        val builder = QueryProductDetailsParams.Product.newBuilder()
        val products: MutableList<QueryProductDetailsParams.Product> = ArrayList()

        // Create a list of Products
        ids.forEach {
            products.add(
                builder
                    .setProductId(it)
                    .setProductType(BillingClient.ProductType.SUBS)
                    .build()
            )
        }

        // Create query
        val params = QueryProductDetailsParams.newBuilder()
            .setProductList(products)
            .build()

        // Async request
        client.queryProductDetailsAsync(params) {
                result: BillingResult,
                list: List<ProductDetails?> ->

            if (result.responseCode == BillingClient.BillingResponseCode.OK)
            {
                Log.d(Billing.LOG, "Fetched ${list.size} SUBS")

                list.filterNotNull().forEach { sub ->

                    // If shitty country/currency - mark as billing unsupported.
                    if (sub.subscriptionOfferDetails?.any { subOffer ->
                            subOffer.pricingPhases.pricingPhaseList.any {
                               BANNED_CURRENCIES.contains(it.priceCurrencyCode.uppercase())
                           }
                        } == true)
                    {
                        subscribeLocally(Billing.UNSUPPORTED)
                    }

                    Billing.products[sub.productId] = Price(context,sub)
                }
            }
            else
            {
                Log.d(Billing.LOG, "Failed to fetch ${list.size} SUBS!")
            }

            Billing.isInitialized.postValue(true)
        }
    }

    private fun querySubsPurchases() {

        // Create query
        val params = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.SUBS)
            .build()

        // Async request
        client.queryPurchasesAsync(params) { result, list ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK)
            {
                acknowledgePurchases(list)
                Log.d(Billing.LOG, "Purchases are acknowledged!")
            }
            else
            {
                Log.d(Billing.LOG, "Failed to query purchases: ${result.responseCode} | ${result.debugMessage}")
            }
        }
    }

    private fun acknowledgePurchases(list: List<Purchase>): Boolean {
        var purchasedSub: String? = null
        list.forEach { purchase ->
            if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED)
            {
                if (!purchase.isAcknowledged)
                {
                    val params = AcknowledgePurchaseParams.newBuilder()
                        .setPurchaseToken(purchase.purchaseToken)
                        .build()
                    client.acknowledgePurchase(params) { }
                }
                purchasedSub = purchase.products[0]
            }
        }
        subscribeLocally(purchasedSub)
        return purchasedSub != null
    }

    fun launchPurchaseFlow(activity: Activity?, sub: Price) {

        if (!client.isReady || activity == null) {
            onPurchase.onError()
            return
        }

        try
        {
            val params: MutableList<ProductDetailsParams> = ArrayList()
            params.add(
                ProductDetailsParams.newBuilder()
                    .setProductDetails(sub.product!!)
                    .setOfferToken(sub.token!!)
                    .build()
            )


            var flowParamsBuilder = BillingFlowParams.newBuilder()
                .setProductDetailsParamsList(params)

            Billing.firebaseAnalyticsID?.let { it ->
                flowParamsBuilder = flowParamsBuilder.setObfuscatedAccountId(it)
            }

            client.launchBillingFlow(activity, flowParamsBuilder.build())
        }
        catch ( _ : IllegalArgumentException)
        {
            onPurchase.onError()
        }
    }
}